torch/csrc/LazyTensor.cpp

```diff
@@ -0,0 +1,385 @@
+#ifndef C10_MOBILE
+
+#include <ATen/core/LazyTensor.h>
+#include <torch/csrc/jit/graph_executor.h>
+#include <torch/csrc/jit/interpreter.h>
+#include <torch/csrc/jit/operator.h>
+#include <torch/csrc/jit/script/function_schema_parser.h>
+
+
+/* LazyTensor Design
+ *
+ *
+ * LazyTensor is an abstraction that enables lazy execution in PyTorch.  It does so by
+ * providing a virtual tensor implementation that looks and feels like a regular Tensor
+ * but does not necessarily execute any of the underlying computation immediately.
+ *
+ * Instead of executing the computation "eagerly,"  LazyTensors will record requested
+ * operations into an implicit graph and later lower that graph to TorchScript IR.
+ * Once `to_eager()` is called, it is guaranteed that any computation required to
+ * resolve the tensor has been completed.
+ *
+ * For backends that digest TorchScript IR, this design is interesting because it enables
+ * these backends to be used in PyTorch without the need for a TorchScript frontend.
+ *
+ * Note: shape propagation is not supported! This means `to_eager` is required before
+ * any shape dependent code is run.  The implication is that some core modules,
+ * such as nn.BatchNorm, cannot be executed lazily because their implementations
+ * check shapes in Python.
+ *
+ * Note: LazyTensor also accumulates a graph during calls to `backward`.  Shape
+ * information is required for error checking during backward propagation,
+ * but with LazyTensors we skip these checks.
+ *
+ * Note: LazyTensor assumes all operations are re-computable!  This assumption
+ * saves us from using too much memory without explicit need.  However, it also
+ * induces potentially unwanted re-compute and BREAKS semantics for operations
+ * that require state.  TODO: blacklist broken ops
+ *
+ * Note: LazyTensors cannot be used in multi-threaded environment.  This paradigm
+ * should subvert the need for parallelism at this level anyway.
+ *
+ */
+
+namespace at {
+
+// @returns pair of output and required stack inputs
+std::pair<torch::jit::Node*, std::vector<const LazyTensorImpl*>> getRequiredGraphValues(
```

Comment:
I'm still not sure what this function does. Perhaps a more elaborated comment would be useful?

Reply:
Having read the function now, the bare minimum description would be to say that this function is what implements lowering to TorchScript IR. But there are some nontrivial choices in how we go about the lowering which I feel are important to also discuss in English.

The model here seems to be that we do a BFS on the implicit graph until we reach non lazy tensors or resolved lazy tensors. Only *resolved* lazy tensors are considered as "inputs" to the graph in question; this is kind of weird, because it seems like there are two ways to end up with a resolved tensor: you called `x.to_lazy()` (this seems appropriate) or you called `y.to_eager()` and overwrote the lazy tensor with the result of the computation (this doesn't seem appropriate).